pandas/core/missing.py

```diff
@@ -681,68 +666,121 @@ def _interp_limit(invalid, fw_limit, bw_limit):
 
-    if bw_limit is not None:
+    we convert the boolean array to integer array for counting the streaks
 
-        if bw_limit == 0:
-            # then we don't even need to care about backwards
-            # just use forwards
-            return f_idx
-        else:
-            b_idx = list(inner(invalid[::-1], bw_limit))
-            b_idx = set(N - 1 - np.asarray(b_idx))
-            if fw_limit == 0:
-                return b_idx
+    >>> arr = arr.astype(int)
+    >>> arr
+    array([1, 1, 0, 0, 1, 1, 1, 0, 1, 1])
 
-    return f_idx & b_idx
+    cumsum will get us off to a good start, we store this as we will need this later
 
+    >>> cumsum = arr.cumsum()
+    >>> cumsum
+    array([1, 2, 2, 2, 3, 4, 5, 5, 6, 7], dtype=int32)
 
-def _rolling_window(a, window):
-    """
-    [True, True, False, True, False], 2 ->
+    multiplying this accumulation with the original array of ones to get non-zero
+    values where we originally had ones
 
-    [
-        [True,  True],
-        [True, False],
-        [False, True],
-        [True, False],
-    ]
+    >>> arr = cumsum * arr
+    >>> arr
+    array([1, 2, 0, 0, 3, 4, 5, 0, 6, 7])
+
+    the previous result is close to what we want, but we want to restart
+    each streak at one. start by using the diff method to substract the previous
+    value for each element
+
+    >>> arr = np.diff(arr, prepend=0)
+    >>> arr
+    array([ 1,  1, -2,  0,  3,  1,  1, -5,  6,  1])
+
+    a negative value now represents the end of a streak of missing values
+    so let's first select just the negative values
+
+    >>> arr = np.where(arr < 0, arr, 0)
+    >>> arr
+    array([ 0,  0, -2,  0,  0,  0,  0, -5,  0,  0])
+
+    we will need to propegate the negative values
+
+    >>> arr = np.minimum.accumulate(arr)
+    >>> arr
+    array([ 0,  0, -2, -2, -2, -2, -2, -5, -5, -5], dtype=int32)
+
+    and then subtract the excess accumlation
+
+    >>> arr = arr + cumsum
+    >>> arr
+    array([1, 2, 0, 0, 1, 2, 3, 0, 1, 2], dtype=int32)
+
+    remember that positive values represent missing values and zeros represent
+    valid values. We have a array with some missing values at the start. For a
+    forward fill algorithm, we want to update the mask to leave these missing
+    values unchanged.
+
+    >>> arr[: arr.argmin()] = 0
+    >>> arr
+    array([0, 0, 0, 0, 1, 2, 3, 0, 1, 2], dtype=int32)
+
+    we will now select only values within a set limit, say 2
+
+    >>> arr = np.where(arr > 2, 0, arr)
+    >>> arr
+    array([0, 0, 0, 0, 1, 2, 0, 0, 1, 2], dtype=int32)
+
+    and finally convert back into a boolean mask
+
+    >>> arr.astype(bool)
+    array([ False,  False, False, False,  True,  True, False, False,  True,
+            True])
     """
-    # https://stackoverflow.com/a/6811241
-    shape = a.shape[:-1] + (a.shape[-1] - window + 1, window)
-    strides = a.strides + (a.strides[-1],)
-    return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)
+
+    def inner(arr, limit):
```

Comment:
will move this into a top level function called ffill_mask_with_limit (in this PR)

and plan to take the max_gap logic from #25141 and have a analogous ffill_mask_with_max_gap (separate PR)